src/animation/state_machine_instance.cpp

```diff
@@ -211,9 +216,72 @@ namespace rive {
             return static_cast<AnimationStateInstance*>(m_CurrentState)->animationInstance();
         }
     };
+
+    /// Representation of a Shape from the Artboard Instance and all the events it
+    /// triggers. Allows tracking hover and performing hit detection only once on
+    /// shapes that trigger multiple events.
+    class HitShape {
+    private:
+        Shape* m_Shape;
+
+    public:
+        Shape* shape() const { return m_Shape; }
+        HitShape(Shape* shape) : m_Shape(shape) {}
+        bool isHovered = false;
+        std::vector<const StateMachineEvent*> events;
+    };
 } // namespace rive
 
-StateMachineInstance::StateMachineInstance(const StateMachine* machine) : m_Machine(machine) {
+void StateMachineInstance::processEvent(Vec2D position, EventType hitEvent) {
+    const int hitRadius = 2;
+    auto hitArea = AABB(position.x() - hitRadius,
+                        position.y() - hitRadius,
+                        position.x() + hitRadius,
+                        position.y() + hitRadius)
+                       .round();
+
+    for (auto hitShape : m_HitShapes) {
+
+        // TODO: quick reject.
+
+        bool isOver = hitShape->shape()->hitTest(hitArea);
+
+        bool hoverChange = hitShape->isHovered != isOver;
+        hitShape->isHovered = isOver;
+
+        // iterate all events associated with this hit shape
+        for (auto event : hitShape->events) {
+            // Always update hover states regardless of which specific event type
+            // we're trying to trigger.
+            if (hoverChange) {
+                if (isOver && event->eventType() == EventType::enter) {
+                    event->performChanges(this);
+                    markNeedsAdvance();
+                } else if (!isOver && event->eventType() == EventType::exit) {
+                    event->performChanges(this);
+                    markNeedsAdvance();
+                }
+            }
+            if (isOver && hitEvent == event->eventType()) {
```

Comment:
Is this mean to handle all *other* eventTypes except enter and exit? (thinking of the previous checks)